### PR TITLE
Publish packages during release pipeline

### PR DESCRIPTION
**Pull request by jamiebrynes7**
This PR introduces a script and a docker container that allows us to publish packages during CI. The idea is that a pipeline would call `./ci/publish-packages.sh`.

This would then:

- Get a copy of the `gdk-for-unity`.
- Get our secrets and throw them in a temp folder.
- Construct a docker image with the required dependencies (`npm`, `jq`, `cloudsmith-cli`, and `spatial`) for running `init.sh` and pushing the packages.
- Run this docker image which will hit init (👊🔴) and then publish each package to Cloudsmith.
- The bash trap runs and deletes our secrets

This PR requires https://github.com/spatialos/gdk-for-unity/pull/1158

**Remaining work**

- [x] Store the cloudsmith API key as a secret
- [x] Figure out how to test this in a safe way (`cloudsmith-cli` has a `--dry-run` mode, could switch based off an env-var and use that?)

### CODE REVIEW

File: ci/docker/publish-packages.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    && apt-get update \
+    && apt-get upgrade -y \
+    && apt-get install -y jq \
+    && curl -LSs -o /usr/local/bin/gosu -SL "https://github.com/tianon/gosu/releases/download/1.4/gosu-$(dpkg --print-architecture)" \
+    && chmod +x /usr/local/bin/gosu \
+    && curl -sL https://deb.nodesource.com/setup_9.x | bash - \
+    && apt-get install -y nodejs \
+    && curl -L https://www.npmjs.com/install.sh | sh
+
+
```
Comment (paulbalaji) [inline ci/docker/publish-packages.Dockerfile:17]: newline

File: ci/publish-packages.sh [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        imp-ci secrets read \
+            --environment="production" \
+            --secret-type="spatialos-service-account" \
+            --buildkite-org="improbable" \
+            --secret-name="unity-gdk-toolbelt" \
+            --field="token" \
+            --write-to="${SPATIAL_OAUTH_FILE}"
+
+
```
Comment (paulbalaji) [inline ci/publish-packages.sh]: newline

File: ci/docker/publish-packages.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,37 @@
+FROM python:3
+
+# Install: 
+#   - cloudsmith-cli            for publishing the packages.
+#   - jq                        for parsing JSON (used in ./init.sh)
+#   - gosu                      for handling issues with spatial writing files which buildkite-agent cannot delete.
+#   - npm                       for packing up our packages
+RUN pip install cloudsmith-cli \
+    && apt-get update \
+    && apt-get upgrade -y \
```
Comment (zeroZshadow) [inline ci/docker/publish-packages.Dockerfile]: Dont run upgrade.
Run `apt-get clean` to decrease image size

File: ci/publish-packages.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+cd "$(dirname "$0")/../"
+
+if [[ -d "${GDK_DIR}" ]]; then
+    rm -rf "${GDK_DIR}"
+fi
+
+mkdir "${GDK_DIR}"
+
+pushd "${GDK_DIR}"
+    git init
```
Comment (zeroZshadow) [inline ci/publish-packages.sh]: This can use `git clone`

File: ci/publish-packages.sh [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    git remote add origin git@github.com:spatialos/gdk-for-unity.git
+    git fetch --depth 1 origin develop
+    git checkout FETCH_HEAD
+popd
+
+getSecrets
+
+docker build --tag local:gdk-publish-packages --file ./ci/docker/publish-packages.Dockerfile .
+docker run -it -v "${SPATIAL_OAUTH_DIR}":/var/spatial_oauth -v "${CLOUDSMITH_AUTH_DIR}":/var/cloudsmith_credentials local:gdk-publish-packages
```
Comment (zeroZshadow) [inline ci/publish-packages.sh]: eof newline

File: ci/docker/publish-packages.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    && apt-get install -y nodejs \
+    && curl -L https://www.npmjs.com/install.sh | sh \
+    && apt-get clean
+
+# Download spatial CLI into container
+ARG TOOLBELT_VERSION="20190827.085359.7e083741fd"
+WORKDIR /build/tools/
+ADD "https://console.improbable.io/toolbelt/download/${TOOLBELT_VERSION}/linux" ./spatial
+RUN ["chmod", "+x", "./spatial"]
+ENV PATH "$PATH:/build/tools/"
```
Comment (zeroZshadow) [inline ci/docker/publish-packages.Dockerfile]:
```suggestion
ENV PATH "${PATH}:/build/tools/"
```

**Review by zeroZshadow — APPROVED** [review]

**Review by jessicafalk — APPROVED** [review]

**Pull request merged**